espnet2/gan_svs/espnet_model.py

```diff
@@ -0,0 +1,1100 @@
+        beat_lab: Optional[torch.Tensor] = None,
+        beat_lab_lengths: Optional[torch.Tensor] = None,
+        beat_xml: Optional[torch.Tensor] = None,
+        beat_xml_lengths: Optional[torch.Tensor] = None,
+        energy: Optional[torch.Tensor] = None,
+        energy_lengths: Optional[torch.Tensor] = None,
+        spembs: Optional[torch.Tensor] = None,
+        sids: Optional[torch.Tensor] = None,
+        lids: Optional[torch.Tensor] = None,
+        flag_IsValid=False,
+        forward_generator: bool = True,
+        **kwargs,
+    ) -> Dict[str, Any]:
+        """Return generator or discriminator loss with dict format.
+
+        Args:
+            text (Tensor): Text index tensor (B, T_text).
+            text_lengths (Tensor): Text length tensor (B,).
+            singing (Tensor): Singing waveform tensor (B, T_wav).
+            singing_lengths (Tensor): Singing length tensor (B,).
+            label_lab (Optional[Tensor]): Label tensor. - phone id sequence
+            label_lab_lengths (Optional[Tensor]): Label length tensor (B,).
+            label_xml (Optional[Tensor]): Label tensor. - phone id sequence
+            label_xml_lengths (Optional[Tensor]): Label length tensor (B,).
+            midi_lab (Optional[Tensor]): Midi tensor.
+            midi_lab_lengths (Optional[Tensor]): Midi length tensor (B,).
+            midi_xml (Optional[Tensor]): Midi tensor.
+            midi_xml_lengths (Optional[Tensor]): Midi length tensor (B,).
+            pitch (Optional[Tensor]): Pitch tensor.
+            pitch_lengths (Optional[Tensor]): Pitch length tensor (B,).
+            tempo_lab (Optional[Tensor]): Tempo tensor.
+            tempo_lab_lengths (Optional[Tensor]): Tempo length tensor (B,).
+            tempo_xml (Optional[Tensor]): Tempo tensor.
+            tempo_xml_lengths (Optional[Tensor]): Tempo length tensor (B,).
+            beat_lab (Optional[Tensor]): Beat tensor.
+            beat_lab_lengths (Optional[Tensor]): Beat length tensor (B,).
+            beat_xml (Optional[Tensor]): Beat tensor.
+            beat_xml_lengths (Optional[Tensor]): Beat length tensor (B,).
+            energy (Optional[Tensor]): Energy tensor.
+            energy_lengths (Optional[Tensor]): Energy length tensor (B,).
+            spembs (Optional[Tensor]): Speaker embedding tensor (B, D).
+            sids (Optional[Tensor]): Speaker ID tensor (B, 1).
+            lids (Optional[Tensor]): Language ID tensor (B, 1).
+            forward_generator (bool): Whether to forward generator.
+            kwargs: "utt_id" is among the input.
+
+        Returns:
+            Dict[str, Any]:
+                - loss (Tensor): Loss scalar tensor.
+                - stats (Dict[str, float]): Statistics to be monitored.
+                - weight (Tensor): Weight tensor to summarize losses.
+                - optim_idx (int): Optimizer index (0 for G and 1 for D).
+
+        """
+        with autocast(False):
+            # Extract features
```

Comment:
I think this part is very similar to svs.py, can we somehow pack the feature extract to a function and reuse the code?